README.md
# prom-metrics-check

`prom-metrics-check` is a command line tools which helps checking metrics between dashboards of grafana and prometheus metrics.

If you download external dashboards to your grafana instance (eg. https://github.com/kubernetes-monitoring/kubernetes-mixin)
you are not 100% sure that all queries used in dashboard work with your prometheus instance. This tool looking for in any
dashboard all queries, then it extracts only metrics and check that all used metrics exist in your prometheus instance.

* Free software: MIT license
* Repository: https://github.com/ContainerSolutions/prom-metrics-check/
* DockerHub: containersol/prom-metrics-check:latest

## Features

* Connect via API to Grafana and get all metrics which are use in any dashboards
* Connect via API to Prometheus and check exist metrics


## Installation

### Stable release

To install prom-metrics-check, run this command directly from repository:


    $ pip install git+https://github.com/ContainerSolutions/prom-metrics-check.git


This is the preferred method to install prom-metrics-check, as it will always install the most recent stable release.

If you don't have [pip] installed, this [Python installation guide] can guide
you through the process.

[pip]: https://pip.pypa.io
[Python installation guide]: http://docs.python-guide.org/en/latest/starting/installation/


### From sources

The sources for prom-metrics-check can be downloaded from the [Github repo].

You can either clone the public repository:


    $ git clone git://github.com/ContainerSolutions/prom-metrics-check

Or download the [tarball]:


    $ curl -OJL https://github.com/ContainerSolutions/prom-metrics-check/tarball/master

Once you have a copy of the source, you can install it with:


    $ python setup.py install

[Github repo]: https://github.com/ContainerSolutions/prom-metrics-check
[tarball]: https://github.com/ContainerSolutions/prom-metrics-check/tarball/master


## Usage

To use this tool locally, you will need to create port-forwards for your grafana and prometheus services.


    $ kubectl port-forward svc/grafana 3000:80
    $ kubectl port-forward svc/prometheus 9090:9090

Also, if you use an API key for Grafana you should set it in you environment.


    $ export GRAFANA_KEY=...

Now you should be able to run the following script:


    $ prom-metrics-check

To see more arguments, run the script with the `--help` flag.


    $ prom-metrics-check --help

## Work with Docker

### Build Docker Image

If you want to build the Docker image, run: `make docker-build`

### Run Docker

To run this container locally, use the `--net=host` flag.

The Makefile can also help running Docker images.

If you download this repository on Linux, you should be able to run the container with: 

    ARGS="--grafana-url=http://localhost:3000 --prometheus-urls=http://localhost:9090 --grafana-key=xyz=" make docker-run

on MacOS the command is slightly differnt:

    ARGS="--grafana-url=http://host.docker.internal:3000 --prometheus-urls=http://host.docker.internal:9090 --grafana-key=xyz=" make docker-run

If you need to run Docker from scratch you should use a different command.

#### Linux

    $ docker run --net=host --rm -e GRAFANA_URL=http://localhost:3000 -e PROMETHEUS_URLS=http://localhost:9090 -e GRAFANA_KEY=xyz containersol/prom-metrics-check:latest

#### MacOS

    $ docker run --net=host --rm -e GRAFANA_URL=http://host.docker.internal:3000 -e PROMETHEUS_URLS=http://host.docker.internal:9090 -e GRAFANA_KEY=xyz containersol/prom-metrics-check:latest


HISTORY.md
# History

### 0.1.0 (2021-11-26)

* Fixed dashboards API call to allow the tool to work with Grafana versions 6.x and newer
(thanks @maciejmazur10c).
* Dropped Python3.5 support. It should still work, we just don't test it there anymore.

### 0.0.1 (2020-07-14)

* First release on PyPI.


tests/__init__.py
"""Unit test package for prom_metrics_check."""


prom_metrics_check/prom_metrics_check.py
from io import BytesIO
from tokenize import tokenize, NUMBER, STRING, NAME, ENCODING, ENDMARKER, \
    NEWLINE

import json
import logging
import urllib.request

logging.basicConfig(level='ERROR')
logger = logging.getLogger('prom-metric-check')


class Token:

    def __init__(self, no, toknum, tokval):
        self.no = no
        self.toknum = toknum
        self.tokval = tokval

    def __repr__(self):
        return 'Token({no}, {toknum}, "{tokval}")'.format(
            no=self.no, toknum=self.toknum, tokval=self.tokval)

    def is_name(self):
        return self.toknum == NAME

    def is_string(self):
        return self.toknum == STRING

    def is_number(self):
        return self.toknum == NUMBER

    def is_operation(self):
        return self.tokval in [
            "+", "-", "*", "/", "%", "^",
            "==", "!=", ">", "<", ">=", "<="]

    def is_colon(self):
        return self.tokval == ":"

    def is_leftbracket(self):
        return self.tokval == "("

    def is_rightbracket(self):
        return self.tokval == ")"

    def is_leftcurltbracket(self):
        return self.tokval == "{"

    def is_rightcurltbracket(self):
        return self.tokval == "}"

    def is_leftsquarebracket(self):
        return self.tokval == "["

    def is_rightsquarebracket(self):
        return self.tokval == "]"

    def is_unnecessary(self):
        return self.tokval in [
            'by', 'without',
            'group_left', 'group_right',
            'and', 'or', 'unless', 'ignoring', 'on',
            'count_values', 'quantile', 'topk', 'bottomk']

    def get_next(self, heap):
        try:
            return heap[self.no + 1]
        except IndexError:
            return None

    def get_prev(self, heap):
        try:
            return heap[self.no - 1]
        except IndexError:
            return None


def tokenize_string(query):
    result, x = [], 0
    g = tokenize(BytesIO(query.encode('utf-8')).readline)
    for toknum, tokval, _a, _b, _c in g:
        if toknum not in [ENCODING, ENDMARKER, NEWLINE]:
            result.append(Token(x, toknum, tokval))
            x += 1
    return result


def find_metrics(tokenized_query):
    skip, temp, heap, metrics = False, None, [], []
    for token in tokenized_query:
        if skip == '(' and token.is_rightbracket():
            skip = False
            if heap:
                if heap[-1].is_leftbracket():
                    heap.pop()
                    if heap and heap[-1].is_name():
                        heap.pop()
                if temp:
                    metrics.append(temp)
                    temp = None
        elif skip in ['{', '['] and (
                token.is_rightcurltbracket() or token.is_rightsquarebracket()):
            skip = False
            if temp:
                metrics.append(temp)
                temp = None
        elif skip:
            continue
        elif token.is_leftbracket():
            heap.append(token)
        elif token.is_leftcurltbracket() or token.is_leftsquarebracket():
            skip = token.tokval
        elif token.is_colon():
            if temp:
                temp += token.tokval
            else:
                temp = token.tokval
        elif token.is_name():
            if token.is_unnecessary():
                if token.get_next(tokenized_query).is_leftbracket():
                    skip = token.get_next(tokenized_query).tokval
            elif token.get_next(tokenized_query):
                if token.get_next(tokenized_query).is_leftbracket():
                    heap.append(token)
                elif token.get_next(tokenized_query).is_leftcurltbracket() or \
                        token.get_next(
                            tokenized_query).is_leftsquarebracket() or \
                        token.get_next(
                            tokenized_query).is_operation() or \
                        token.get_next(
                            tokenized_query).is_rightbracket():
                    if temp:
                        temp += token.tokval
                        metrics.append(temp)
                        temp = None
                    else:
                        metrics.append(token.tokval)
                elif token.get_next(tokenized_query).is_colon():
                    if not temp:
                        temp = token.tokval
                    else:
                        temp += token.tokval
            elif not token.get_next(tokenized_query):
                if temp:
                    temp += token.tokval
                    metrics.append(temp)
                    temp = None
                else:
                    metrics.append(token.tokval)
    return list(set(metrics))


class Response:
    def __init__(self, response=None):
        self.response = response
        self.content = self.response.read()

    @property
    def ok(self):
        return 200 <= self.response.getcode() <= 299

    @property
    def text(self):
        return self.content.decode('utf-8')

    def json(self):
        if self.text:
            return json.loads(self.text)
        return {}


def request_get(url=None, token=None):
    request = urllib.request.Request(url)
    if token:
        request.add_header(
            "Authorization", "Bearer {token}".format(token=token))
    try:
        conn = urllib.request.urlopen(request, timeout=5)
    except urllib.error.URLError as e:
        print('{url}: {error}'.format(url=url, error=e.reason))
        exit(1)
    return Response(conn)


def check_exist_metrics(except_metrics=None, url=None):
    logger.info("Except metrics: {metrics}".format(
        metrics=', '.join(except_metrics)))
    exist_metrics = request_get(
        '{url}/api/v1/label/__name__/values'.format(url=url))
    if exist_metrics.ok:
        data = list(set(exist_metrics.json()['data']))
        logger.info("Exist metrics: {metrics}".format(metrics=', '.join(data)))
        difference = set(except_metrics).difference(
            set(exist_metrics.json()['data']))
        return list(difference)
    raise ValueError


def get_recursively(search_dict, field):
    """
    Takes a dict with nested lists and dicts, and searches all dicts
    for a key of the field provided.
    """
    fields_found = []

    for key, value in search_dict.items():

        if key == field:
            fields_found.append(value)

        elif isinstance(value, dict):
            results = get_recursively(value, field)
            for result in results:
                fields_found.append(result)

        elif isinstance(value, list):
            for item in value:
                if isinstance(item, dict):
                    more_results = get_recursively(item, field)
                    for another_result in more_results:
                        fields_found.append(another_result)

    return fields_found


def get_all_metrics(dashboards=None):
    expression, metrics = [], []
    for configration in dashboards:
        expression.extend(get_recursively(configration, 'expr'))
    for expr in expression:
        metrics.extend(find_metrics(tokenized_query=tokenize_string(expr)))
    return list(set(metrics))


def load_dashboard(url=None, key=None):
    dashboards, dashboards_uid = [], []
    search_dashboard = request_get(
        '{url}/api/search'.format(url=url), token=key)
    if search_dashboard.ok:
        dashboards_uid = [dash.get('uid') for dash in search_dashboard.json()]
    for name in dashboards_uid:
        dashboard = request_get('{url}/api/dashboards/uid/{name}'.format(
            url=url, name=name), token=key)
        if dashboard.ok:
            dashboards.append(dashboard.json().get('dashboard', {}))
    return dashboards


prom_metrics_check/cli.py
"""Console script for prom-metrics-check."""
import os
import sys
import argparse
import logging

from .prom_metrics_check import load_dashboard, get_all_metrics, \
    check_exist_metrics


logging.basicConfig(level='ERROR')
logger = logging.getLogger('prom-metric-check')


GRAFANA_DEFAULT_URL = 'http://localhost:3000'
PROMETHEUS_DEFAULT_URL = 'http://localhost:9090'

GRAFANA_URL = os.environ.get(
    'GRAFANA_URL', GRAFANA_DEFAULT_URL)
GRAFANA_KEY = os.environ.get(
    'GRAFANA_KEY')
PROMETHEUS_URLS = os.environ.get(
    'PROMETHEUS_URLS', PROMETHEUS_DEFAULT_URL).split(',')


def main(args=None):
    """Console script for prom_metrics_check."""
    parser = argparse.ArgumentParser(
        description='Command line tool for check metrics between '
                    'grafana and prometheus instance.')
    parser.add_argument(
        '--grafana-url',
        metavar='grafana_url',
        help='Set grafana url. Default value is {url}'.format(
            url=GRAFANA_DEFAULT_URL),
        nargs='?',
        default=GRAFANA_URL)
    parser.add_argument(
        '--grafana-key',
        metavar='grafana_key',
        help='Set grafana key to have API access.',
        nargs='?',
        default=GRAFANA_KEY)
    parser.add_argument(
        '--prometheus-urls',
        metavar='prometheus_urls',
        help='Set prometheus url. Default value is {url}'.format(
            url=PROMETHEUS_DEFAULT_URL),
        nargs='*',
        default=PROMETHEUS_URLS)
    args = parser.parse_args(args)
    for url in args.prometheus_urls:
        dashboards = load_dashboard(url=args.grafana_url, key=args.grafana_key)
        except_metrics = get_all_metrics(dashboards=dashboards)
        missing_metrics = check_exist_metrics(except_metrics, url)
        if missing_metrics:
            logger.critical(
                " Metrics which don't exist: {metrics}".format(
                    metrics=', '.join(missing_metrics)))
    return 0


if __name__ == '__main__':
    sys.exit(main())    # pragma: no cover
